File: 35-search-insert-position/35-search-insert-position.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var searchInsert = function(nums, target) {
  const search = (l, r) => {
    if (nums[l] >= target) return l
    if (nums[r] === target) return r
    if (nums[r] < target) return r + 1
    const mid = l + (~~((r-l) / 2))
    if (nums[mid] == target) return mid
    else if (nums[mid] < target) return search(mid + 1, r)
    else return search(l, mid - 1)
  }
  return search(0, nums.length - 1)
};

File: 35-search-insert-position/35-search-insert-position.go
func searchInsert(nums []int, target int) int {
  l, r := 0, len(nums)
  for l < r {
    m := l + (r-l)/2
    if nums[m] == target { return m }
    if nums[m] < target { l = m + 1}
    if nums[m] > target { r = m }
  }
  return l
}